server/src-lib/Hasura/GraphQL/Resolve/Mutation.hs

```diff
@@ -68,49 +69,65 @@ convertRowObj val =
-  (CAUpdate, Just cons) -> return $ RI.CP1Update (RI.Constraint cons) cols
+mkConflictClause :: RI.ConflictCtx -> RI.ConflictClauseP1
+mkConflictClause (RI.CCDoNothing constrM) =
+  RI.CP1DoNothing $ fmap RI.Constraint constrM
+mkConflictClause (RI.CCUpdate constr updCols) =
+  RI.CP1Update (RI.Constraint constr) updCols
 
 parseAction
   :: (MonadError QErr m)
-  => AnnGObject -> m ConflictAction
-parseAction obj = do
-  val <- onNothing (Map.lookup "action" obj) $ throw500
-    "\"action\" field is expected but not found"
-  (enumTy, enumVal) <- asEnumVal val
-  withPathK "action" $ case G.unName $ G.unEnumValue enumVal of
-    "ignore" -> return CAIgnore
-    "update" -> return CAUpdate
-    _ -> throw500 $ "only \"ignore\" and \"updated\" allowed for enum type " <> showNamedTy enumTy
+  => AnnGObject -> m (Maybe ConflictAction)
+parseAction obj =
+  mapM parseVal $ Map.lookup "action" obj
+  where
+    parseVal val = do
+      (enumTy, enumVal) <- asEnumVal val
+      withPathK "action" $ case G.unName $ G.unEnumValue enumVal of
+        "ignore" -> return CAIgnore
+        "update" -> return CAUpdate
+        _ -> throw500 $
+          "only \"ignore\" and \"updated\" allowed for enum type "
+          <> showNamedTy enumTy
 
 parseConstraint
   :: (MonadError QErr m)
-  => AnnGObject -> m (Maybe ConstraintName)
+  => AnnGObject -> m ConstraintName
 parseConstraint obj = do
-  t <- mapM parseVal $ Map.lookup "constraint" obj
-  return $ fmap ConstraintName t
+  v <- onNothing (Map.lookup "constraint" obj) $ throwVE
+    "\"constraint\" is expected, but not found"
+  parseVal v
   where
     parseVal v = do
       (_, enumVal) <- asEnumVal v
-      return $ G.unName $ G.unEnumValue enumVal
+      return $ ConstraintName $ G.unName $ G.unEnumValue enumVal
+
+parseUpdCols
+  :: (MonadError QErr m)
+  => AnnGObject -> m (Maybe [PGCol])
+parseUpdCols obj =
+  mapM parseVal $ Map.lookup "update_columns" obj
+  where
+    parseVal val = flip withArray val $ \_ enumVals ->
+      forM enumVals $ \eVal -> do
+        (_, v) <- asEnumVal eVal
+        return $ PGCol $ G.unName $ G.unEnumValue v
 
 parseOnConflict
   :: (MonadError QErr m)
-  => AnnGValue -> m RI.ConflictCtx
-parseOnConflict val =
+  => [PGCol] -> AnnGValue -> m RI.ConflictCtx
+parseOnConflict inpCols val =
   flip withObject val $ \_ obj -> do
-    action <- parseAction obj
-    constraintM <- parseConstraint obj
-    return (action, constraintM)
+    actionM <- parseAction obj
+    constraint <- parseConstraint obj
+    updColsM <- parseUpdCols obj
+    -- consider "action" if "update_columns" is not mentioned
+    return $ case (updColsM, actionM) of
+      (Just [], _)             -> RI.CCDoNothing $ Just constraint
+      (Just cols, _)           -> RI.CCUpdate constraint cols
+      (Nothing, Just CAIgnore) -> RI.CCDoNothing $ Just constraint
+      (Nothing, Just CAUpdate) -> RI.CCUpdate constraint inpCols
+      (Nothing, Nothing)       -> RI.CCDoNothing $ Just constraint
```

Comment:
this should also be same as the previous case, when `update_columns` is missing, we take it that they want to update the specified columns